hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelinePlacementPolicy.java

```diff
@@ -21,6 +21,7 @@
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hdds.conf.OzoneConfiguration;
```

Comment:
There are a lot of changes in this test file, but do we have a test that reproduces the problem reported? The problem is that if the anchor node is on a rack with only a single node, then only 2 nodes would be allocated rather than 3? It should be possible to reproduce this easily with a 3 node and 3 rack cluster, one rack per node.

I think we should add a test that reproduces the problem so we know the changes correct it and we have no regressions in the future. We should also test this via the public API of PipelinePlacementPolicy if possible. Right now, most of the tests in TestPipelinePolicy seem to call "@VisibleForTesting" methods rather than chooseDatanodes or getResultSet, which are the public methods.

Reply:
I just added a test for this case. Please see the updates.